tests/rptest/tests/storage_timing_stress_test.py

```diff
@@ -0,0 +1,339 @@
+        if not consumer_complete:
+            return False, f"Consumer consumed only {consumed} out of {produced} messages"
+
+        return True, ""
+
+    def is_complete(self):
+        complete, reason = self._check_completion()
+        if complete:
+            return True
+
+        delta = datetime.now() - self.test_start_ts
+        max_runtime = self.target_runtime * self.allow_runtime_overshoot_by
+        if delta.total_seconds() > max_runtime:
+            raise TimeoutError(
+                f"Workload did not complete within {max_runtime}s: {reason}")
+
+        return False
+
+    def prologue(self, cleanup_policy):
+        self.topic_spec.cleanup_policy = cleanup_policy
+        self.topics = [self.topic_spec]
+        self._create_initial_topics()
+
+        self.producer = self._create_producer(cleanup_policy)
+        self.consumer = self._create_consumer(self.producer)
+
+        self.producer.start()
+
+        # Sleep for a bit to hit the cloud storage read path when consuming
+        time.sleep(3)
+        self.consumer.start()
+
+        self.test_start_ts = datetime.now()
+
+    def epilogue(self, cleanup_policy):
+        self.producer.wait()
+        self.consumer.wait()
+
+    def register_check(self, name, check_fn):
+        self.checks.append(StorageCheck(name, check_fn))
+
+    def do_checks(self):
+        if len(self.checks) == 0:
+            return
+
+        with ThreadPoolExecutor(max_workers=len(self.checks)) as executor:
+
+            def start_check(check):
+                self.logger.info(f"Check {check.name} starting")
+                return executor.submit(check.check, self)
+
+            futs = {start_check(check): check for check in self.checks}
+
+            done, not_done = concurrent.futures.wait(
+                futs, timeout=self.check_interval)
+
+            failure_count = 0
+            for f in done:
+                check_name = futs[f].name
+                if ex := f.exception():
+                    self.logger.error(
+                        f"Check {check_name} threw an exception: {ex}")
+                    failure_count += 1
+                else:
+                    self.logger.info(
+                        f"Check {check_name} completed successfully")
+
+            for f in not_done:
+                check_name = futs[f].name
+                self.logger.error(
+                    f"Check {check_name} did not complete within the check interval"
+                )
+
+            if failure_count > 0 or len(not_done) > 0:
+                raise RuntimeError(
+                    f"Failed checks: {failure_count}; Incomplete checks: {len(not_done)}"
+                )
+
+            self.logger.info(f"All checks completed successfully")
+
+    @cluster(num_nodes=5)
+    @parametrize(cleanup_policy="delete")
+    @parametrize(cleanup_policy="compact,delete")
+    @skip_debug_mode
+    def test_storage(self, cleanup_policy):
```

Comment:
It would be nice to add in some periodic `DeleteRecords` calls. We've recently seen the two interacting and stopping local retention from being applied.

Reply:
doing that now